### PR TITLE
Listen sys call comments + tests

### PR DESCRIPTION
**Pull request by davidge20**
## Description

Added comments to listen_syscall and added two tests

Fixes # (issue)

 

Two issues that I would like to share:
1. While calling the listen sys call, the backlog is set to 5 in the function, so the test queues more than 5 connections. The 6th connection should return with an error, given the queue is maximized. However, the 6th queue returns with EINPROGRESS.
2. We create a blocking socket to connect to a "server" and it returns immediately. The blocking socket should not return with success immediately as the server did not accept the connection.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- Test A - `ut_lind_net_listen_more_than_backlog()`
- Test B - `ut_lind_net_listen_nonblock_then_block()`

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: rustc-ice-2024-07-10T16_06_47-19579.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                               at /rustc/6be96e3865c4e59028fd50396f7a46c3498ce91d/library/alloc/src/boxed.rs:2064:9
+  17:     0xffff7f2971e8 - std::sys::pal::unix::thread::Thread::new::thread_start::h9b0657d0725f835d
+                               at /rustc/6be96e3865c4e59028fd50396f7a46c3498ce91d/library/std/src/sys/pal/unix/thread.rs:108:17
+  18:     0xffff7f0d597c - <unknown>
+  19:     0xffff7f13ba4c - <unknown>
+  20:                0x0 - <unknown>
+
+
+rustc version: 1.81.0-nightly (6be96e386 2024-07-09)
+platform: aarch64-unknown-linux-gnu
```
Comment (qianxichen233) [inline rustc-ice-2024-07-10T16_06_47-19579.txt]: could you remove these redundent files? You can either uncommit this commit with `git reset --soft HEAD^` and commit again with these files removed, or just create a new commit with these files removed

File: src/safeposix/syscalls/net_calls.rs [hunk 2]
```diff
@@ -1510,22 +1510,81 @@ impl Cage {
...
+    /// ### Description
+    /// 
+    /// `listen_syscall` listen for connections on a socket
+    /// 
+    /// ### Arguments
+    /// 
+    /// it accepts two parameters: 
+    /// * `sockfd` - a file descriptor that refers to a socket
+    ///    of type SOCK_STREAM or SOCK_SEQPACKET.
+    /// // ** Do we deal with SOCK_SEQPACKET ?? ** //
```
Comment (qianxichen233) [inline src/safeposix/syscalls/net_calls.rs]: I don't think we support SOCK_SEQPACKET. You can mention this somewhere in the comment

File: src/safeposix/syscalls/net_calls.rs [hunk 3]
```diff
@@ -1510,22 +1510,81 @@ impl Cage {
...
+    ///   that the socket isn't assigned an address
+    /// * unknown errno value from socket listen sys call from libc
+    /// 
+    /// for more detailed description of all the commands and return values, see 
+    /// [listen(2)](https://linux.die.net/man/2/listen)
+    //
+    // TODO: We are currently ignoring backlog 
     pub fn listen_syscall(&self, fd: i32, _backlog: i32) -> i32 {
+        //If fd is out of range of [0,MAXFD], process will panic
+        //Otherwise, we obtain a write gaurd to the Option<FileDescriptor> object
```
Comment (qianxichen233) [inline src/safeposix/syscalls/net_calls.rs:1885]: I think you should either mark a BUG here, or just fix it (return EBADF if fd is out of range)

File: src/safeposix/syscalls/net_calls.rs [hunk 4]
```diff
@@ -1510,22 +1510,81 @@ impl Cage {
...
+                    //success
                     match sockhandle.state {
                         ConnState::LISTEN => {
-                            return 0; //Already done!
+                            return 0;
                         }
 
+                        //If the given socket is connected to another socket or 
+                        //if a non blocking socket is in progress of connecting
+                        //to another socket
```
Comment (qianxichen233) [inline src/safeposix/syscalls/net_calls.rs]: I think you can add more comments explaining each state below. Like what each state is, and what does that state mean to a socket

File: src/safeposix/syscalls/net_calls.rs [hunk 5]
```diff
@@ -1563,18 +1629,27 @@ impl Cage {
...
+                            //to readily accept connections
                             NET_METADATA.listening_port_set.insert(porttuple.clone());
                             sockhandle.state = ConnState::LISTEN;
 
-                            let listenret = sockhandle.innersocket.as_ref().unwrap().listen(5); //default backlog in repy for whatever reason, we replicate it
+                            //Call listen from libc on the socket
+                            //Set the backlog to 5:
+                            //default backlog in repy for whatever reason, we replicate it
+                            //** Would we ever want to change the backlog??  **/
+                            let listenret = sockhandle.innersocket.as_ref().unwrap().listen(5);
```
Comment (qianxichen233) [inline src/safeposix/syscalls/net_calls.rs]: But we do have a backlog argument passed into listen_syscall, why not just use it directly?

File: src/safeposix/syscalls/net_calls.rs [hunk 6]
```diff
@@ -1586,30 +1661,44 @@ impl Cage {
...
                                 ));
+                                //Set the socket state to NOTCONNECTED, as
+                                //the socket is not listening
                                 sockhandle.state = ConnState::NOTCONNECTED;
                                 return lr;
                             };
 
-                            //set rawfd for select
+                            //set rawfd for select sys call
+                            //** Why is this being done in listen ??
```
Comment (qianxichen233) [inline src/safeposix/syscalls/net_calls.rs]: I guess since there is a code logic before that does something like "if the socket didn't bind to any address, then assign a new address". So the innersocket could be possibly a new one that just created in this syscall, and we certainly want to update its rawfd as well

File: src/tests/networking_tests.rs [hunk 7]
```diff
@@ -2642,6 +2783,7 @@ pub mod net_tests {
     }
 
     #[test]
+    #[ignore]
```
Comment (qianxichen233) [inline src/tests/networking_tests.rs]: why this is marked as ignore?

File: src/tests/networking_tests.rs [hunk 8]
```diff
@@ -669,6 +670,146 @@ pub mod net_tests {
...
+        assert_eq!(cage.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        lindrustfinalize();
+    }
+
+    //We create a blocking socket to connect to the server and it returns immediately
+    //The blocking socket should not return with success immediately as the server
+    //did not accept the connection
+    #[test]
+    #[ignore]
+    pub fn ut_lind_net_listen_nonblock_then_block() {
```
Comment (qianxichen233) [inline src/tests/networking_tests.rs]: I don't quite understand this test, what's the issue here? Like the connect is supposed to block forever since server is not accepting the socket, but it actually returned immediately?

File: rustc-ice-2024-07-10T16_02_42-17210.txt [hunk 9]
```diff
@@ -0,0 +1,36 @@
+thread 'coordinator' panicked at /rustc/6be96e3865c4e59028fd50396f7a46c3498ce91d/compiler/rustc_codegen_ssa/src/back/write.rs:1676:29:
+/rustc/6be96e3865c4e59028fd50396f7a46c3498ce91d/compiler/rustc_codegen_ssa/src/back/write.rs:1676:29: worker thread panicked
```
Comment (namanlalitnyu) [inline rustc-ice-2024-07-10T16_02_42-17210.txt:2]: I think we should remove these `rustc-ice*` files from these changes.

File: src/safeposix/syscalls/net_calls.rs [hunk 10]
```diff
@@ -1546,12 +1609,15 @@ impl Cage {
...
                             if sockhandle.domain == AF_UNIX {
                                 sockhandle.state = ConnState::LISTEN;
                                 return 0;
                             }
 
+                            //If the given socket is not assigned an address,
+                            //attempt to bind the socket to an address.
                             if sockhandle.localaddr.is_none() {
                                 let shd = sockhandle.domain as i32;
                                 let ibindret = self._implicit_bind(&mut *sockhandle, shd);
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/net_calls.rs:1958]: Can you explain a bit about this function through a couple of comments for better understanding.

File: rustc-ice-2024-07-10T16_06_47-19579.txt (hunk 1 above)
Comment (davidge20) [reply rustc-ice-2024-07-10T16_06_47-19579.txt]: Perfect, thanks!

Had to use `git reset --mixed HEAD^` so the changes arent commited again automatically and need to be added with git add.

File: src/safeposix/syscalls/net_calls.rs (hunk 5 above)
Comment (davidge20) [reply src/safeposix/syscalls/net_calls.rs]: Good question. @rennergade why was the backlog set to 5 and not the number passed in to listen?

File: src/tests/networking_tests.rs (hunk 7 above)
Comment (davidge20) [reply src/tests/networking_tests.rs]: The test fails given it's not giving the behavior we expect. Right now if one test fails all of the following tests fail. Once I think yash fixes that we don't need to ignore it anymore.

File: src/tests/networking_tests.rs (hunk 8 above)
Comment (davidge20) [reply src/tests/networking_tests.rs]: Aren't INET sockets are blocking by default? Hence, as you mentioned, shouldn't the connect sys call block forever since the server is not accepting the socket? However, it returns with success immediately.

File: src/tests/networking_tests.rs (hunk 7 above)
Comment (namanlalitnyu) [reply src/tests/networking_tests.rs]: Can you try it now with the latest rebase, looks like Yash's Changes are in related to this.

File: src/tests/networking_tests.rs (hunk 8 above)
Comment (qianxichen233) [reply src/tests/networking_tests.rs]: From what I've read, connect is actually not going to block if server already listened but hasn't accept the connection yet. `connect` can return normally if it goes into the server's listening queue successfully.
btw another suggestion is that it might not be a good idea to add the test that would expect `blocking forever` behavior

Comment (davidge20) [reply src/tests/networking_tests.rs]: Yeah thats a good point about adding a test that expects blocking forever behavior. 

Based on the following: https://stackoverflow.com/questions/8258107/does-connect-block-for-tcp-socket , 
so even if the TCP socket is in blocking mode, connect will return with success by being put on the server's listening queue cause the OS takes care of that? 

Then whats the point of blocking mode?

Comment (qianxichen233) [reply src/tests/networking_tests.rs]: nonblocking mode could affect many other functions like accept, send, recv, etc. And in case of connect, there is still some thing get affected. For example, when you try to connect in nonblocking mode, if the server cannot respond immediately (note that this is different from server refuse to respond, in which case the server would explictly send a message back (TCP RST specifically I think)), connect could block until it receives something from server. But in nonblocking mode, it will return after a few milliseconds (must wait some time to determine if server can respond because of network delay) if the server is not responding with INPROGRESS state. Later when the connection is either denied or accept, you can check it using getsockopt syscall. I think the blocking/nonblocking behavior on `connect` syscall, especially on INET is very complex, mostly because you need to balance network delay and nonblocking delay, and it may involve a lot of TCP stuff

**Review by qianxichen233 — APPROVED** [review]

**Comment by davidge20** [comment]
@rupeshkoushik07  requesting your review, thanks!

**Review by namanlalitnyu — APPROVED** [review]
LGTM!

**Review by rupeshkoushik07 — APPROVED** [review]
LGTM !

File: src/tests/networking_tests.rs [hunk 11]
```diff
@@ -2761,6 +2857,7 @@ pub mod net_tests {
     }
 
     #[test]
+    #[ignore]
```
Comment (yashaswi2000) [inline src/tests/networking_tests.rs]: whats the current behaviour?

File: src/safeposix/syscalls/net_calls.rs (hunk 5 above)
Comment (yashaswi2000) [reply src/safeposix/syscalls/net_calls.rs]: bumping this up @rennergade

**Review by yashaswi2000 — APPROVED** [review]
LGTM

File: src/safeposix/syscalls/net_calls.rs (hunk 5 above)
Comment (rennergade) [reply src/safeposix/syscalls/net_calls.rs]: Yeah i dont see any reason to not use the passed parameter. Lets make that change. Just a weird legacy thing.

File: src/safeposix/syscalls/net_calls.rs [hunk 12]
```diff
@@ -1865,12 +1935,23 @@ impl Cage {
                                 );
                             }
 
-                            // simple if it's a domain socket
+                            //If the given socket is a Unix socket, lind handles
+                            //the connection, return with success
                             if sockhandle.domain == AF_UNIX {
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs:1941]: I think if were going to actually add backlog in for INET its worth putting a TODO tag here saying implement backlog for unix

File: src/tests/networking_tests.rs (hunk 11 above)
Comment (rennergade) [reply src/tests/networking_tests.rs]: yeah this one definitely shouldnt fail so we need to look into it

File: src/tests/networking_tests.rs [hunk 13]
```diff
@@ -2,6 +2,7 @@
 pub mod net_tests {
     use super::super::*;
     use crate::interface;
+    use crate::interface::errnos::{syscall_error, Errno};
```
Comment (rennergade) [inline src/tests/networking_tests.rs]: why did you need to import this?

**Review by rennergade — CHANGES_REQUESTED** [review]
This is very close and quite thorough. Good job!

A few minor changes, lets add the backlog and make sure that test is working.

File: src/tests/networking_tests.rs (hunk 11 above)
Comment (davidge20) [reply src/tests/networking_tests.rs]: some mistake on my end, removed the ignore statement

File: src/tests/networking_tests.rs (hunk 13 above)
Comment (davidge20) [reply src/tests/networking_tests.rs]: Wasn't necessary, i figured out how to use errno

removed the statement

**Review by rennergade — APPROVED** [review]
